src/storage/cache/rpc/packet.rs

```diff
@@ -0,0 +1,525 @@
+
+    /// Set timestamp
+    fn set_timestamp(&mut self, timestamp: u64);
+    /// Get timestamp
+    fn get_timestamp(&self) -> u64;
+
+    /// Serialize request data to bytes
+    fn set_req_data(&mut self, data: &[u8]) -> Result<(), RpcError<String>>;
+
+    /// Get the serialized request data
+    fn get_req_data(&self) -> Result<Vec<u8>, RpcError<String>>;
+
+    /// Serialize response data to bytes
+    fn set_resp_data(&mut self, data: &[u8]) -> Result<(), RpcError<String>>;
+
+    /// Deserialize response data from bytes
+    fn get_resp_data(&self) -> Result<Vec<u8>, RpcError<String>>;
+
+    /// Get the packet status
+    fn status(&self) -> PacketStatus;
+
+    /// Set the packet status
+    fn set_status(&mut self, status: PacketStatus);
+}
+
+/// The `PacketStatus` enum is used to represent the status of a packet.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PacketStatus {
+    /// The packet is pending.
+    Pending,
+    /// The packet is successful.
+    Success,
+    /// The packet is failed.
+    Failed,
+    /// The packet is timeout.
+    Timeout,
+}
+
+impl PacketStatus {
+    /// Convert the `PacketStatus` to u8
+    #[must_use]
+    pub fn to_u8(&self) -> u8 {
+        match *self {
+            PacketStatus::Pending => 0,
+            PacketStatus::Success => 1,
+            PacketStatus::Failed => 2,
+            PacketStatus::Timeout => 3,
+        }
+    }
+
+    /// Convert u8 to `PacketStatus`
+    #[must_use]
+    pub fn from_u8(status: u8) -> Self {
+        match status {
+            0 => PacketStatus::Pending,
+            1 => PacketStatus::Success,
+            3 => PacketStatus::Timeout,
+            _ => PacketStatus::Failed,
+        }
+    }
+}
+
+/// The `PacketsInner` struct is used to store the current tasks.
+#[derive(Debug)]
+struct PacketsInner<P>
+where
+    P: Packet + Send + Sync,
+{
+    /// current tasks, marked by the seq number
+    packets: HashMap<u64, P>,
+}
+
+impl<P: Packet + Send + Sync> PacketsInner<P> {
+    /// Create a new `PacketsInner`
+    #[must_use]
+    pub fn new() -> Self {
+        PacketsInner {
+            packets: HashMap::new(),
+        }
+    }
+
+    /// Add a task to the packets
+    pub fn add_task(&mut self, seq: u64, packet: P) {
+        self.packets.insert(seq, packet);
+    }
+
+    /// Get a task from the packets and remove it
+    pub fn remove_task(&mut self, seq: u64) -> Option<P> {
+        // remove packet and return it
+        self.packets.remove(&seq)
+    }
+
+    /// Clean pending tasks as timeout
+    /// In first step, we will try to mark the task as timeout if it is pending and timeout
+    /// In 10 * timeout time, we will force to remove the task if the task is not consumed
+    pub fn clean_timeout_tasks(&mut self, current_timestamp: u64, timeout: u64) {
+        let mut last_timeout_packets = Vec::new();
+        for (seq, packet) in &mut self.packets {
+            let current_status = packet.status();
```

Comment:
This function has logical problem, it should be one step. ignore the PacketStatus

Reply:
This was originally designed to be a delayed deletion, now the packet notifies the caller immediately that the behaviour has been deleted.